src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java

```diff
@@ -67,121 +94,489 @@ public Future<String> handle(KafkaConsumerRecord<String, String> event) {
-        targetInstanceCollection, kafkaHeaders));
+      })
+      .onFailure(throwable -> {
+        String warningMessage = String.format("Instance with InstanceId=%s is not exists on target tenant: %s.", instanceId, targetTenant);
+        LOGGER.info(warningMessage);
+      })
+      .recover(throwable -> publishInstance(sharingInstanceMetadata, sourceInstanceCollection, targetInstanceCollection, kafkaHeaders));
   }
 
+
   private Future<String> publishInstance(SharingInstance sharingInstanceMetadata, InstanceCollection sourceInstanceCollection,
                                          InstanceCollection targetInstanceCollection, Map<String, String> kafkaHeaders) {
 
-    String instanceId = sharingInstanceMetadata.getInstanceIdentifier().toString();
-    String sourceTenant = sharingInstanceMetadata.getSourceTenantId();
-    String targetTenant = sharingInstanceMetadata.getTargetTenantId();
+    LOGGER.info("publishInstance :: Publishing instance with InstanceId={} from tenant={} to tenant={}.",
+      sharingInstanceMetadata.getInstanceIdentifier(), sharingInstanceMetadata.getSourceTenantId(),
+      sharingInstanceMetadata.getTargetTenantId());
 
-    return getInstanceById(instanceId, sourceTenant, sourceInstanceCollection)
-      .compose(srcInstance -> {
-        if ("FOLIO".equals(srcInstance.getSource())) {
-          JsonObject jsonInstance = srcInstance.getJsonForStorage();
-          jsonInstance.remove(HRID_KEY);
-          return addInstance(srcInstance, targetTenant, targetInstanceCollection)
-            .compose(addedInstance -> {
-              JsonObject jsonInstanceToPublish = srcInstance.getJsonForStorage();
-              jsonInstanceToPublish.put("source", CONSORTIUM_FOLIO.getValue());
-              return updateInstanceInStorage(Instance.fromJson(jsonInstanceToPublish), sourceInstanceCollection)
-                .map(ignored -> "Instance has been updated successfully")
-                .onSuccess(ignored -> {
-                  String message = format("Instance with InstanceId=%s has been shared to the target tenant %s",
-                    instanceId, targetTenant);
-                  sendCompleteEventToKafka(sharingInstanceMetadata, COMPLETE, message, kafkaHeaders);
-                }).onFailure(throwable -> {
-                  String errorMessage = format("Error updating Instance with InstanceId=%s on source tenant %s.",
-                    instanceId, sourceTenant);
-                  sendErrorResponseAndPrintLogMessage(errorMessage, sharingInstanceMetadata, kafkaHeaders);
-                });
+    Promise<String> promise = Promise.promise();
+
+    try {
+      String instanceId = sharingInstanceMetadata.getInstanceIdentifier().toString();
+      String sourceTenant = sharingInstanceMetadata.getSourceTenantId();
+      String targetTenant = sharingInstanceMetadata.getTargetTenantId();
+
+      getInstanceById(instanceId, sourceTenant, sourceInstanceCollection).onSuccess(srcInstance -> {
```

Comment:
In a previous PR, I had mention that this chain has too much nesting. Need to break them out into private methods. There shouldn't be more than three levels of indentation.